File: examples/csharp/Patterns/Adapter.cs
namespace DesignPatterns.Adapter
{
    // Existing XML service
    public class XmlDataSource
    {
        public string GetXmlData() =>
            "<data><name>Alice</name><age>30</age></data>";
    }

    // Target interface the client expects
    public interface IJsonDataSource
    {
        string GetJsonData();
    }

    // Adapter
    public class XmlToJsonAdapter : IJsonDataSource
    {
        private readonly XmlDataSource _xmlSource;

        public XmlToJsonAdapter(XmlDataSource xmlSource) => _xmlSource = xmlSource;

        public string GetJsonData()
        {
            var xml = _xmlSource.GetXmlData();
            // Simplified conversion for demonstration
            return xml
                .Replace("<data>", "{ ")
                .Replace("</data>", " }")
                .Replace("<name>", "\"name\": \"")
                .Replace("</name>", "\", ")
                .Replace("<age>", "\"age\": ")
                .Replace("</age>", "");
        }
    }

    public class AnalyticsClient
    {
        public void ProcessJson(IJsonDataSource source)
        {
            Console.WriteLine($"Processing JSON: {source.GetJsonData()}");
        }
    }

    public static class Demo
    {
        public static void Run()
        {
            var xmlSource = new XmlDataSource();
            Console.WriteLine($"XML source: {xmlSource.GetXmlData()}");

            var adapter = new XmlToJsonAdapter(xmlSource);
            var client = new AnalyticsClient();
            client.ProcessJson(adapter);
        }
    }
}

File: examples/cpp/adapter.cpp
#include <iostream>
#include <string>

class JSONAnalytics {
public:
    void processJSON(const std::string& json) const {
        std::cout << "Processing JSON data: " << json << std::endl;
    }
};

class XMLData {
public:
    std::string getXML() const {
        return "<data><name>Widget</name><price>9.99</price></data>";
    }
};

class XMLToJSONAdapter : public JSONAnalytics {
    XMLData xmlData_;
public:
    XMLToJSONAdapter(const XMLData& data) : xmlData_(data) {}

    void analyze() {
        std::string xml = xmlData_.getXML();
        std::cout << "Adapter: Converting XML to JSON" << std::endl;
        std::cout << "  Input XML: " << xml << std::endl;
        std::string json = R"({"name": "Widget", "price": 9.99})";
        std::cout << "  Converted JSON: " << json << std::endl;
        processJSON(json);
    }
};

int main() {
    std::cout << "=== Adapter: XML to JSON ===" << std::endl;

    XMLData xmlData;
    XMLToJSONAdapter adapter(xmlData);
    adapter.analyze();

    return 0;
}
